src/isomorphic/devtools/ReactComponentTreeDevtool.js

```diff
@@ -0,0 +1,180 @@
+/**
+ * Copyright 2016-present, Facebook, Inc.
+ * All rights reserved.
+ *
+ * This source code is licensed under the BSD-style license found in the
+ * LICENSE file in the root directory of this source tree. An additional grant
+ * of patent rights can be found in the PATENTS file in the same directory.
+ *
+ * @providesModule ReactComponentTreeDevtool
+ */
+
+'use strict';
+
+var invariant = require('invariant');
+
+var unmountedContainerIDs = [];
+var allChildIDsByContainerID = {};
+var tree = {};
+
+function updateTree(id, update) {
+  if (!tree[id]) {
+    tree[id] = {
+      parentID: null,
+      ownerID: null,
+      text: null,
+      childIDs: [],
+      displayName: 'Unknown',
+      isTopLevelWrapper: false,
+    };
+  }
+  update(tree[id]);
+}
+
+function purgeTree(id) {
+  var item = tree[id];
+  if (!item) {
+    return;
+  }
+
+  var {childIDs, containerID} = item;
+  delete tree[id];
+
+  if (containerID) {
+    allChildIDsByContainerID[containerID] = allChildIDsByContainerID[containerID]
+      .filter(childID => childID !== id);
+  }
+  childIDs.forEach(purgeTree);
+}
+
+var ReactComponentTreeDevtool = {
+  onSetIsTopLevelWrapper(id, isTopLevelWrapper) {
+    updateTree(id, item => item.isTopLevelWrapper = isTopLevelWrapper);
+  },
+
+  onSetIsComposite(id, isComposite) {
+    updateTree(id, item => item.isComposite = isComposite);
+  },
+
+  onSetDisplayName(id, displayName) {
+    updateTree(id, item => item.displayName = displayName);
+  },
+
+  onSetChildren(id, nextChildIDs) {
+    if (ReactComponentTreeDevtool.isTopLevelWrapper(id)) {
```

Comment:
This is dangerous. The API contract allows setTopLevelWrapper to be mutated and doesn't say anything about the order of execution. This assumes that it is set before the children. To do this correctly, you'd have to empty the children if setTopLevelWrapper ever gets set after I guess.

I think we can avoid this whole problem though... [see next comment]